### PR TITLE
feat: new flush policy - purge flush

### LINKED ISSUE #68: Implement follower flush
**Description**

`flush` dumps memtable to sst file.

CeresDB current doesn't distinguish behaviors between leader and follower. In follower it only needs to purge its memtables and skip dumping.

**Proposal**

Implement the follower flush behavior described above. Serval steps I come up with:
- [ ] Receive and track the max sequence number flushed by leader
- [ ] Flush (drop) memtable based on the max seq
- [ ] (considering) Flush behavior after becoming the new leader.

**Additional context**

This issue may only cover the implementation part. The trigger will come with a follow up PR, which depends on WAL replication
and/or instance role.

### PR DESCRIPTION
**Pull request by waynexia**
# Which issue does this PR close?

Closes #68 

# Rationale for this change
 

Implement `Purge` flush policy. Detailed in #68.

# What changes are included in this PR?

- new flush option `TableFlushPolicy`
- new field in `TableFlushRequest`: `max_sequence` and `policy`
- change `MemTableView::pick_memtables_to_flush` to pick memtable based on sequence number rather than memtable id

# Are there any user-facing changes?
no

# How does this change test

By existing UT

### CODE REVIEW

**Comment by waynexia** [comment]
Two thoughts:
- deprecate `max_table_id` as `sequence_number` is the thing makes decisions
- consider reclaiming memtables as soon as they get dumped. They are removed together after all the memtables are dumped now.

File: analytic_engine/src/instance/flush_compaction.rs [hunk 1]
```diff
@@ -167,6 +167,19 @@ pub struct TableFlushRequest {
...
+    pub policy: TableFlushPolicy,
+}
+
+/// Policy of how to perform flush operation.
+#[derive(Debug)]
+pub enum TableFlushPolicy {
+    /// Dump memtable to sst file.
+    Dump,
+    /// Drop memtables.
+    Purge,
```
Comment (ShiKaiWi) [inline analytic_engine/src/instance/flush_compaction.rs:182]: I guess we need one policy called `Block` which handles that the follower's memtables can't be flushed until the flushed sequence number from its leader is greater than its memtable to flush.

Comment (waynexia) [reply analytic_engine/src/instance/flush_compaction.rs:182]: This is covered in `pick_memtables_to_flush`. It only picks memtables whose `last_sequence` ≤ sequence number in the flush request.

File: analytic_engine/src/instance/flush_compaction.rs [hunk 2]
```diff
@@ -167,6 +167,21 @@ pub struct TableFlushRequest {
...
     /// Max id of memtable to flush (inclusive).
     pub max_memtable_id: MemTableId,
+    /// Max sequence number to flush (inclusive).
+    pub max_sequence: SequenceNumber,
+    /// Flush policy.
+    pub policy: TableFlushPolicy,
+}
+
+/// Policy of how to perform flush operation.
+#[derive(Debug)]
```
Comment (ShiKaiWi) [inline analytic_engine/src/instance/flush_compaction.rs:177]: Shoult it derive Copy?

File: analytic_engine/src/instance/flush_compaction.rs [hunk 3]
```diff
@@ -167,6 +167,21 @@ pub struct TableFlushRequest {
...
+pub enum TableFlushPolicy {
```
Comment (ShiKaiWi) [inline analytic_engine/src/instance/flush_compaction.rs:178]: Maybe this should be called `TableFlushAction` and the `Policy` may be reserved to for the trait to decide the `TableFlushRequest`, e.g.:
```rust
pub trait TableFlushPolicy {
    fn decide(&self, ...) -> Result<TableFlushRequest>;
}
```

File: analytic_engine/src/instance/flush_compaction.rs [hunk 4]
```diff
@@ -350,40 +371,116 @@ where
...
-    /// Caller should guarantee flush of single table is sequential
-    async fn flush_memtables_to_outputs(&self, flush_req: &TableFlushRequest) -> Result<()> {
-        // TODO(yingwen): Record memtables num to flush as statistics
+    /// Each table can only have one running flush job.
+    async fn flush_memtables(&self, flush_req: &TableFlushRequest) -> Result<()> {
         let TableFlushRequest {
             table_data,
-            max_memtable_id,
+            // TODO: consider deprecate this field,
+            max_memtable_id: _,
```
Comment (ShiKaiWi) [inline analytic_engine/src/instance/flush_compaction.rs:379]: Can it be deprecated just in this pr?

**Review by ShiKaiWi — APPROVED** [review]

**Pull request merged**